rfd/0105-ec2-tags.md

```diff
@@ -0,0 +1,115 @@
+---
+authors: Roman Tkachenko (roman@goteleport.com), Scale Team
+state: draft
+---
+
+# RFD 105 - Importing EC2 tags via API
+
+## Required approvers
+
+Engineering: @fspmarshall, @espadolini, @rosstimothy
+
+## What
+
+Proposes a way to import instance tags using AWS EC2 API instead of metadata
+endpoint during EC2 auto-discovery.
+
+## Why
+
+Teleport nodes running on AWS currently automatically import EC2 instance tags
+as server labels using instance metadata endpoint, as was implemented by the
+[RFD 72](https://github.com/gravitational/teleport/blob/master/rfd/0072-ec2-tags.md).
+
+Some users disable metadata endpoint due to security concerns but still would
+like to have Teleport import server labels from AWS.
+
+This RFD describes a mechanism by which instance labels will instead be fetched
+by the discovery service and reconciled with the node's own labels.
+
+## Scope
+
+This feature will only apply to nodes enrolled in the cluster using auto-discovery
+mechanism. Other nodes, even if they're running on EC2, will keep functioning
+as-is (at least for now), i.e. importing tags from EC2 metadata endpoint if it's
+available.
+
+## How
+
+We will discuss design and implementation of 3 components of the solution:
+
+- Fetching EC2 instance tags.
+- Reconciling tags during heartbeat.
+- Propagating tags to the node.
+
+### Fetching tags
+
+Teleport discovery service is currently responsible for scanning user's AWS
+account on a regular cadence and enrolling matching instances into the cluster
+([RFD 57](https://github.com/gravitational/teleport/blob/master/rfd/0057-automatic-aws-server-discovery.md)).
+It is a service that users run on their own infrastructure.
+
+Discovery service uses `DescribeInstances` API. Right now the returned instance
+list is only used for cluster enrollment (i.e. running appropriate commands on
+them through SSM). With the changes proposed in this RFD, the discovery service
+will convert each EC2 `Instance` into a new helper Teleport resource that we'll
+call `DiscoveredServer`. `DiscoveredServer` will include the instance's tags as
+labels, as well as other supporting information like instance ID, account ID
+and region. Each label key will be prefixed with `aws/`.
+
+After each discovery loop cycle, the discovery service will save all new and
+updated `DiscoveredServer` resources in the cluster backend. Discovery service
+maintains an in-memory cache of all discovered EC2 instances and will only
+upsert resources that are new or had their labels changed.
+
+To reduce load on auth and backend, discovery service will send them in batches
+of 100 using `UpsertDiscoveredServers()` API and stagger batches over some time
+interval (e.g. one batch per second).
```

Comment:
note: agent-side caching will need to take resources expiry into account, as unchanged `DiscoveredServer` resources will still need to be re-emitted on occasion in order to ensure that the resources are preserved.

I'd suggest that we target a longer TTL for this resource than the standard 10 minutes in order to limit the frequency with which writes need to occur.

Also, if we use a fixed batch-size, then there will be a hard upper bound given the batch size and the resource TTL, after which writes will fail to keep up with the total number of instances.  We should either pick a combination that supports a very optimistic upper limit (e.g. 100k), or we should consider batching as a portion of the total number of discovered EC2 instances (e.g. `(len(instances) / 600) + 1` per second to ensure that all instances all have their labels hb once every 10 minutes).